pkg/pubsub/adapter/adapter.go

```diff
@@ -18,253 +18,202 @@ package adapter
-	Name string `envconfig:"NAME" required:"true"`
-
-	// Environment variable containing the resource group. E.g., storages.events.cloud.google.com.
-	ResourceGroup string `envconfig:"RESOURCE_GROUP" default:"pullsubscriptions.pubsub.cloud.google.com" required:"true"`
-
-	// inbound is the cloudevents client to use to receive events.
-	inbound cloudevents.Client
-
-	// outbound is the cloudevents client to use to send events.
-	outbound cloudevents.Client
-
-	// transformer is the cloudevents client to transform received events before sending.
-	transformer cloudevents.Client
+	adapterType string
 
 	// reporter reports metrics to the configured backend.
 	reporter StatsReporter
-}
 
-// Start starts the adapter. Note: Only call once, not thread safe.
-func (a *Adapter) Start(ctx context.Context) error {
-	var err error
+	logger *zap.Logger
+}
 
-	if a.SendMode == "" {
-		a.SendMode = converters.DefaultSendMode
+// NewAdapter creates a new adapter.
+func NewAdapter(
+	ctx context.Context,
+	subscription *pubsub.Subscription,
+	outbound *nethttp.Client,
+	reporter StatsReporter,
+	sinkURI SinkURI,
+	transformerURI TransformerURI,
+	adapterType AdapterType,
+	extensions map[string]string) *Adapter {
+	return &Adapter{
+		subscription:   subscription,
+		outbound:       outbound,
+		reporter:       reporter,
+		sinkURI:        string(sinkURI),
+		transformerURI: string(transformerURI),
+		adapterType:    string(adapterType),
+		extensions:     extensions,
+		logger:         logging.FromContext(ctx),
 	}
+}
 
-	// Convert base64 encoded json map to extensions map.
-	a.extensions, err = utils.Base64ToMap(a.ExtensionsBase64)
-	if err != nil {
-		fmt.Printf("[warn] failed to convert base64 extensions to map: %v", err)
-	}
+func (a *Adapter) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	// Receive Events on Pub/Sub.
-	if a.inbound == nil {
-		if a.inbound, err = a.newPubSubClient(ctx); err != nil {
-			return fmt.Errorf("failed to create inbound cloudevent client: %w", err)
-		}
-	}
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- a.subscription.Receive(ctx, a.receive)
+	}()
 
-	// Send events on HTTP.
-	if a.outbound == nil {
-		if a.outbound, err = a.newHTTPClient(ctx, a.Sink); err != nil {
-			return fmt.Errorf("failed to create outbound cloudevent client: %w", err)
-		}
+	// Stop either if the adapter stops (sending to errCh) or if the context Done channel is closed.
+	select {
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		break
 	}
 
-	if a.reporter == nil {
-		a.reporter = NewStatsReporter()
+	// Done channel has been closed, we need to gracefully shutdown. The cancel() method will start its
+	// shutdown, if it hasn't finished in a reasonable amount of time, just return an error.
+	cancel()
+	select {
+	case err := <-errCh:
+		return err
+	case <-time.After(30 * time.Second):
+		return errors.New("timeout shutting down adapter")
 	}
+}
```

Comment:
I don't quite understand this logic. To reach `cancel()` it seems that either `Receive` returned an error or context was done. If it's an error (1st case), then it means the pull has already stopped. If the context was done (2nd case), then it should already be handled by the `Receive` function. So it looks to me the `cancel()` is redundant and we just need to wait 30s. Thoughts?

Reply:
I think you are right, I copied/pasted this from some eventing stuff. I'll double check.
Will fix the other stuff first, and then address this.

Reply:
Done, thanks Chen!